R/shared-methods.R

```diff
@@ -509,6 +509,53 @@ DataClass <- R6::R6Class(
         source_data_cols = paste(unlist(self$source_data_cols), collapse = ", ")
       )
       return(sum_df)
+    },
+
+    #' @description Specific test for a specific class.
+    #' By default this method is empty as if any code is required it should be
+    #' defined in a child class specific `specific_tests` method.
+    #' @param self_copy R6class the object to test
+    specific_tests = function(self_copy) {
+
+    },
+
+    #' @description Run tests on data class. Inherited by child classes so test
+    #' run through each class. Tests data can be downloaded (if requested),
+    #' cleaned, processed and returned. Tests run on a cloned copy of the target
+    #' class.
+    #' @param download logical. To download the data (TRUE) or use a snapshot
+    #' (FALSE). Defaults to FALSE.
+    #' @param snapshot_path character_array the path to save the downloaded
+    #' @param copy logical To work on a copy of the class or the actual class
+    #' @param ... Additional parameters to pass to `specific_tests`
+    #' snapshot to. Defaults to TRUE.
+    test = function(download = FALSE, snapshot_path = "",
+                    copy = TRUE, ...) {
+      data_name <- paste0(class(self)[1], " at level ", self$level)
+      if (copy) {
+        test_class <- self$clone()
+      } else {
+        test_class <- self
+      }
+      test_download(
+        self = test_class,
+        download = download,
+        data_name = data_name,
+        snapshot_path = snapshot_path
+      )
+      test_cleaning(
+        self = test_class,
+        data_name = data_name
+      )
+      test_processing(
+        self = test_class,
+        data_name = data_name
+      )
+      test_return(
+        self = test_class,
+        data_name = data_name
+      )
+      self$specific_tests(self_copy = test_class, ...)
```

Comment:
if we are keeping specific_tests why not use the if exists eval logic used for dispatch in clean (so specific_test doesn't have to exist if empty)